service/history/tasks/key.go

```diff
@@ -0,0 +1,122 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tasks
+
+import (
+	"fmt"
+	"time"
+)
+
+var (
+	DefaultFireTime         = time.Unix(0, 0)
+	defaultFireTimeUnixNano = DefaultFireTime.UnixNano()
+)
+
+type (
+	Key struct {
+		// FireTime is the scheduled time of the task
+		FireTime time.Time
+		// TaskID is the ID of the task
+		TaskID int64
+	}
+
+	Keys []Key
+)
+
+func NewImmediateKey(taskID int64) Key {
+	return Key{
+		FireTime: DefaultFireTime,
+		TaskID:   taskID,
+	}
+}
+
+func NewKey(fireTime time.Time, taskID int64) Key {
+	return Key{
+		FireTime: fireTime,
```

Comment:
add a sanity check that fireTime.UnixNano() >= 0?

Reply:
yeah I think we discussed that before, but think it's a bit risky and would first validate it on a test cluster.
The change can be found here: https://github.com/temporalio/temporal/tree/panic-invalid-task-key, I should probably rebase it on master.